src/test/disconnected_transactions.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+//
+#include <boost/test/unit_test.hpp>
+#include <core_memusage.h>
+#include <kernel/disconnected_transactions.h>
+#include <test/util/setup_common.h>
+
+BOOST_FIXTURE_TEST_SUITE(disconnected_transactions, TestChain100Setup)
+
+//! Tests that DisconnectedBlockTransactions limits its own memory properly
+BOOST_AUTO_TEST_CASE(disconnectpool_memory_limits)
+{
+    if (sizeof(void*) != 8) {
+        BOOST_TEST_MESSAGE("Exiting disconnected_transactions memory usage test early due to unsupported arch");
+    }
+
+    // Use the coinbase transactions from TestChain100Setup. It doesn't matter whether these
+    // transactions would realistically be in a block together, they just need distinct txids and
+    // uniform size for this test to work.
+    std::vector<CTransactionRef> block_vtx(m_coinbase_txns);
+    BOOST_CHECK_EQUAL(block_vtx.size(), 100);
+
+    // Roughly estimate sizes to sanity check that DisconnectedBlockTransactions::DynamicMemoryUsage
+    // is within an expected range.
+
+    // Overhead for the hashmap depends on number of buckets
+    std::unordered_map<uint256, CTransaction*, SaltedTxidHasher> temp_map;
+    temp_map.reserve(1);
+    const size_t MAP_1{memusage::DynamicUsage(temp_map)};
+    temp_map.reserve(100);
+    const size_t MAP_100{memusage::DynamicUsage(temp_map)};
+
+    const size_t TX_USAGE{RecursiveDynamicUsage(block_vtx.front())};
+    for (const auto& tx : block_vtx)
+        BOOST_CHECK_EQUAL(RecursiveDynamicUsage(tx), TX_USAGE);
+
+    const size_t ENTRY_USAGE_OVERESTIMATE{TX_USAGE + 8 * memusage::MallocUsage(sizeof(void*))};
```

Comment:
The 8 pointers isn't to create an overestimation margin, it's to account for the approximate usage of the data structures (pointers in the list nodes, etc.)

Reply:
Maybe a comment like this would suffice?
```
// Our overall formula is unordered map overhead + usage per entry.
// Implementations may vary, but we're trying to guess the usage of data structures:
// list: per entry x (2 pointers + element itself, which is a CTransactionRef)
// unordered map: hashtable overhead + per entry x (3 pointers + value itself, which is an iterator)
// Usage per entry is thus approximately `TX_USAGE` + 6 pointers. Add another 2 pointers as a safety margin.
```

Reply:
Added the comment but I think, its word to also explain whats the pointers represent also
```
// Our overall formula is unordered map overhead + usage per entry.
// Implementations may vary, but we're trying to guess the usage of data structures:
// list: per entry x (2 pointers (prev, next) + element itself, which is a CTransactionRef)
// unordered map: hashtable overhead + per entry x (3 pointers (key, value, internal structure for book keeping) + value itself, which is an iterator)
// Usage per entry is thus approximately `TX_USAGE` + 6 pointers. Add another 2 pointers as a safety margin.
```
But not so sure if its correct for unordered map, I added as you suggested, if you think the explanation is worth it happy to repush

Reply:
In combination with [the fix suggested here](https://github.com/bitcoin/bitcoin/pull/28385#discussion_r1363659227), what do you think about this?

It makes all the assumptions explicit in code (except for the pointers which are documented in docstring). It also removes the safety margin. I think we can leave that out until/if we observe it causing issues, e.g. on certain platforms?

<details>
<summary>git diff on 8faa9800891c8b3eaef051ad9f0dc4d61475cebe</summary>

```diff
diff --git a/src/test/disconnected_transactions.cpp b/src/test/disconnected_transactions.cpp
index b488866a37..bae9850de9 100644
--- a/src/test/disconnected_transactions.cpp
+++ b/src/test/disconnected_transactions.cpp
@@ -34,10 +34,14 @@ BOOST_AUTO_TEST_CASE(disconnectpool_memory_limits)
 
     // Our overall formula is unordered map overhead + usage per entry.
     // Implementations may vary, but we're trying to guess the usage of data structures:
-    // list: per entry x (2 pointers (next pointer and prev pointer) + element itself, which is a CTransactionRef)
-    // unordered map: hashtable overhead + per entry x (3 pointers + value itself, which is an iterator)
-    // Usage per entry is thus approximately `TX_USAGE` + 6 pointers. Add another 2 pointers as a safety margin.
-    const size_t ENTRY_USAGE_OVERESTIMATE{TX_USAGE + 8 * memusage::MallocUsage(sizeof(void*))};
+    const size_t ENTRY_USAGE_OVERESTIMATE{
+        TX_USAGE
+        // list entry: 2 pointers (next pointer and prev pointer) + element itself
+        + memusage::MallocUsage((2 * sizeof(void*)) + sizeof(decltype(block_vtx)::value_type))
+        // unordered map: 1 pointer for the hashtable + key and value
+        + memusage::MallocUsage(sizeof(void*) + sizeof(decltype(temp_map)::key_type)
+                                + sizeof(decltype(temp_map)::value_type))
+    };
 
     // DisconnectedBlockTransactions that's just big enough for 1 transaction.
     {

```
<details>